### DISCUSSION TITLE
How to get the partial derivatives from a CombiTable2D

### DISCUSSION THREAD

**Discussion by franko** [OpenModelica/OpenModelica#12379] [Q&A] [answered]
Hi everyone,

I've been trying to find a solution by searching the internet and looking at the official Modelica documentation, but I haven't had much success so far.

I'm working on a simple task: I'm using a CombiTable2D to describe a variable z as a function of two variables, x and y, like this:

z = f(x, y)

What I need is to obtain the partial derivatives of z with respect to x and y, so that I can use them in the equations defining my model. However, I haven't found a straightforward way to achieve this.

I tried using a Modelica function, as the documentation mentions that partial derivatives work with functions, but I couldn't figure out how to define a function that uses a CombiTable2D, OpenModelica refused to accept a Function that use a CombiTable2D as a protected object instance.

I also looked into the internals of CombiTable2D and found the getDerTable2DValue function, which seems related to derivatives, but I couldn't determine how to use it for my purpose.

As a workaround, I'm currently computing the derivatives myself using Excel (!) and writing them into two separate tables, let's call them dfdx and dfdy. I then save these tables in a .txt file and load them using two separate CombiTable2D objects, one for each derivative.

I would greatly appreciate any help or guidance on how to properly obtain the partial derivatives of a CombiTable2D in Modelica. Thank you in advance!

**Comment by casella** [comment]
Unfortunately the partial function derivative operator can be applied to functions, not to blocks. I'm afraid that would be tricky

**Reply by franko** [reply, marked as answer]
Thank you for having a look at that.

Actually I got an answer on stackoverflow that works:

https://stackoverflow.com/questions/78440503/how-to-obtain-partial-derivatives-of-a-combitable2d-in-modelica

I didn't assess completely the accuracy of the derivatives but it seems to work fine. I guess the derivatives are conform to the specification according to the smoothness parameter.

**Reply by casella** [reply]
Nice 😃 

Could you please open a ticket on https://github.com/modelica/ModelicaStandardLibrary/issues, suggesting that we add those output as conditionally activated outputs? That would make it usable much more easily.
